### DISCUSSION TITLE
Replace history state without redirection

### DISCUSSION THREAD

**Discussion by bubokko** [remix-run/react-router#11415] [Proposals]
As far as I know, currently it's not possible to replace the history state without redirection. It's a bit surprising, given it's super useful to put data to be remembered on the current page in case user navigates away and then comes back. Even the native `history.replaceState()` puts the `state` param first in order to allow state change without changing the URL. The docs for `react-router` discourage using `history.state` directly, while not providing support for its basic features. I'm confused.

In my opinion, it should also cause stable behavior, like `useRef()`. It's just a piece of data attached to the current context. If that triggers more `useEffect()` and re-renders, it would be annoying to work with.

The most obvious scenario is remembering scrolling positions of certain scrollable elements (not just the `<body>`), but there are many, many more use cases.

**Comment by gerdemb** [comment]
I agree. As a workaround, I wrote this helper function which uses the History API directly to replace state. Note, that the state provided by the react-router `useLocation()` hook is actually stored in the `usr` property of `history.state`.

```
export default function replaceState(state: any) {
  history.replaceState(
    {
      ...history.state,
      usr: state,
    },
    ""
  );
}
```

**Reply by bubokko** [reply]
Are you 100% sure all React Router's history state is in under `usr` key? Can't check now, but I remember that at some point I have changed the history state directly and it somehow messed up with the React Router.

**Reply by gustavopch** [reply]
Yes, you can verify here: https://github.com/remix-run/react-router/blob/89049ce4ee1b265b78468f25eb494eef240a7e40/packages/router/history.ts

Edit: it's not that "all React Router's history state is under `usr`", the `usr` property only holds the part defined by the user. There are also `idx` and `key` that are managed by React Router.

**Reply by bubokko** [reply]
I mean, if I update history state silently like that, won't it somehow screw up the flow of React Router? I would consider keeping my own entries under a different property because of that. Am I going too far with that?

**Reply by gerdemb** [reply]
I would say the workaround relies on an "undocumented API" which could possibly change without notice. However, it works for now. Would be nice if a properly supported solution was provided in the future....

**Reply by superlucky84** [reply]
It seems like the safest method at the moment.
I think it would be useful if it was provided as an official API.
